add sizehint! for first and make append!/prepend! safer

vtjnash opened the pull request:
First we add an optional API parameter for `sizehint!` that controls whether it is for `push!` (default) or `pushfirst!`. Secondly, we make the offset zero when using `sizehint!` to shrink an array from the end, or the trailing size zero when using it to shring from the beginning.

Then we replace the prior implementations of `prepend!` and `append!` with ones that are safe even if the iterator changes length during the operation or if convert fails. The result of `prepend!` may be in an undefined order (because of the `reverse!` call) in the presence of concurrent modifications or errors, but at least all of the elements will be present and valid afterwards.

Replaces and closes #49905
Replaces and closes #47391
Fixes #15868

Benchmarks show that repeated `push!` performance (with sizehint) is nearly equivalent to the old append performance:
```
julia> @benchmark append!(x, 1:1000) setup=x=Vector{Float64}(undef,0)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.027 μs … 72.871 μs  ┊ GC (min … max): 0.00% … 94.57%
 Time  (median):     1.465 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.663 μs ±  1.832 μs  ┊ GC (mean ± σ):  6.20% ±  5.67%

             ▂▃▅▆█▇▇▆▄▂▁                                      
  ▂▁▁▂▂▂▂▃▄▅▇█████████████▇▆▅▅▅▅▅▅▄▅▄▅▅▅▆▇███▆▅▄▃▃▂▂▂▂▂▂▂▂▂▂ ▄
  1.03 μs        Histogram: frequency by time        2.31 μs <

 Memory estimate: 19.69 KiB, allocs estimate: 0.

julia> @benchmark append!(x, 1:1000) setup=x=Vector{Int}(undef,0)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  851.900 ns … 76.757 μs  ┊ GC (min … max): 0.00% … 91.59%
 Time  (median):       1.181 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.543 μs ±  1.972 μs  ┊ GC (mean ± σ):  6.75% ±  5.75%

     ▆█▇▃                                                       
  ▂▃██████▇▅▅▄▅▅▃▂▂▂▃▃▃▂▃▃▃▂▂▂▂▂▁▂▁▂▁▂▂▂▁▁▂▂▁▁▁▁▁▁▁▂▂▂▃▃▃▃▂▂▂▂ ▃
  852 ns          Histogram: frequency by time         4.07 μs <

 Memory estimate: 19.69 KiB, allocs estimate: 0.
```

oscardssmith commented on base/array.jl:
```diff
@@ -1441,13 +1436,16 @@ function resize!(a::Vector, nl::Integer)
 
 """
-    sizehint!(s, n) -> s
+    sizehint!(s, n, [first=false]; shink=false) -> s
```
```suggestion
    sizehint!(s, n, [first=false]; shink=true) -> s
```

barucden replied:
watch out for the typo in `shink`

vtjnash commented:
@nanosoldier `runbenchmarks("array", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51903#issuecomment-1786189483) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d64df4_vs_96147bb/report.md).

rfourquet commented on base/array.jl:
```diff
@@ -1441,13 +1436,16 @@ function resize!(a::Vector, nl::Integer)
 the cost of incremental reallocation by doing it once up front; this can improve
 performance.
 
+If `first` is true, then the reserved space is from the start of the
+collection, if `pushfirst!` is supported on the collection.
```
It could be interesting to widen the scope of the `first` argument, e.g. `sizehint!(::BitSet)` could also benefit from this argument (and just forward it to the underlying vector in the current implementation), but does not support `pushfirst!`.

vtjnash replied:
Good point. I changed the documentation to mention ordered collections, so that `pushfirst` is now just an example instead of the spec

vtjnash committed:
add sizehint! for first and make append!/prepend! safer

First we add an optional API parameter for `sizehint!` that controls
whether it is for `push!` (default) or `pushfirst!`. Secondly, we make
the offset zero when using `sizehint!` to shrink an array from the end,
or the trailing size zero when using it to shring from the beginning.

Then we replace the prior implementations of `prepend!` and `append!`
with ones that are more safe, even if the iterator changes length during
the operation or if convert fails. The result of `prepend!` may be in an
undefined order (because of the `reverse!` call) in the presence of
concurrent modifications or errors, but at least all of the elements
will be present and all entries will be valid afterwards.

Replaces and closes #49905
Replaces and closes #47391
Fixes #15868

Co-authored-by: Sukera <Seelengrab@users.noreply.github.com>
Co-authored-by: MasonProtter <mason.protter@icloud.com>
